File: 016/cpp/main.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        std::sort(nums.begin(), nums.end());

        int r = nums[0] + nums[1] + nums[2] - target;
        for (int i = 0; i < nums.size(); i++) {
            int t = nums[i];
            int j = i + 1;
            int k = nums.size() - 1;
            while (j < k) {
                int t0 = t + nums[j] + nums[k] - target;
                if (std::abs(t0) < std::abs(r)) {
                    r = t0;
                }
                if (t0 < 0) {
                    j++;
                } else if (t0 > 0) {
                    k--;
                } else {
                    return target;
                }
            } 
        }
        return r + target;
    }
};

File: 016/rust/main.rs
impl Solution {
    pub fn three_sum_closest(nums: Vec<i32>, target: i32) -> i32 {
        let mut nums = nums;
        nums.sort();

        let mut r = nums[0] + nums[1] + nums[2] - target;
        for i in 0..nums.len() {
            let t = nums[i];
            let mut j = i + 1;
            let mut k = nums.len() - 1;
            while j < k {
                let t0 = t + nums[j] + nums[k] - target;
                if t0.abs() < r.abs() {
                    r = t0;
                }                
                if t0 < 0 {
                    j += 1;
                } else if t0 > 0 {
                    k -= 1;
                } else {
                    return target;
                }
            }
        }
        return r + target;

    }
}
